cluster/v1alpha1/helpers.go

```diff
@@ -0,0 +1,311 @@
+package v1alpha1
+
+import (
+	"fmt"
+	"math"
+	"regexp"
+	"sort"
+	"strconv"
+	"strings"
+	"time"
+
+	"k8s.io/apimachinery/pkg/util/intstr"
+	"k8s.io/apimachinery/pkg/util/sets"
+	"k8s.io/utils/clock"
+	clusterv1beta1 "open-cluster-management.io/api/cluster/v1beta1"
+)
+
+var RolloutClock = clock.Clock(clock.RealClock{})
+var maxTimeDuration = time.Duration(math.MaxInt64)
+
+type RolloutStatus int
+
+const (
+	// resource desired status is not applied yet
+	ToApply RolloutStatus = iota
+	// resource desired status is applied and last applied status is not updated
+	Progressing
+	// resource desired status is applied and last applied status is succeed
+	Succeed
+	// resource desired status is applied and last applied status is failed
+	Failed
+)
+
+// Return the rollout status on each managed cluster
+type ClusterRolloutStatusFunc func(clusterName string) ClusterRolloutStatus
+
+type ClusterRolloutStatus struct {
+	// rollout status
+	Status RolloutStatus
+	// the last transition time of rollout status
+	LastTransitionTime *time.Time
+}
+
+// +k8s:deepcopy-gen=false
+type RolloutHandler struct {
+	// placement decision tracker
+	pdTracker *clusterv1beta1.PlacementDecisionClustersTracker
+}
+
+func NewRolloutHandler(pdTracker *clusterv1beta1.PlacementDecisionClustersTracker) *RolloutHandler {
+	if pdTracker == nil {
+		return nil
+	}
+
+	return &RolloutHandler{
+		pdTracker: pdTracker,
+	}
+}
+
+// The input is a duck type RolloutStrategy and a ClusterRolloutStatusFunc to return the rollout status on each managed cluster.
+// Return the strategy actual take effect and a list of cluster that needs to update with current state.
+func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, map[string]ClusterRolloutStatus, error) {
```

Comment:
It makes sense I think. Can have GroupKey added, only for the clusters need to update eg,
```
type ClusterRolloutStatus struct {
        // cluster group name and group key
        GroupKey GroupKey
	// rollout status
	Status RolloutStatus
	// the last transition time of rollout status
	LastTransitionTime *time.Time
}

func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, map[string]ClusterRolloutStatus, error) {...
```
Or maybe a whole picture, eg,
```
type ClustersRolloutStatus map[string]ClusterRolloutStatus

type RolloutStatus struct {
	MandatoryGroupsSucceed bool
       // progressive with mandatory groups or progressive per group will return this field
        DecisionGroupsStatus map[GroupKey]ClustersRolloutStatus 
        // all or progressive non mandatory group clusters will return in this field
	ClustersStatus ClustersRolloutStatus 
        // clusters to rollout, all the rollout strategy will return this field
	ClustersToRollout ClustersRolloutStatus
}

func (r *RolloutHandler) GetRolloutCluster(rolloutStrategy RolloutStrategy, statusFunc ClusterRolloutStatusFunc) (*RolloutStrategy, RolloutStatus, error) {...

```
can discuss more in this part.

Reply:
Per the offline discussion, the returned ClusterRolloutStatus will contain the group name and group index. 

Also, since the timeout clusters are intended to skip and continue to rollout others, the returned value would be like below, separate clusters to rollout and clusters that are timeout. 
```
type RolloutResult struct {
	// clusters to rollout
	ClustersToRollout map[string]ClusterRolloutStatus
	// clusters that are timeout
	ClustersTimeOut map[string]ClusterRolloutStatus
}
```